File: src/AltinnCore/ServiceLibrary/Models/UserProfile.cs
using AltinnCore.ServiceLibrary.Enums;

namespace AltinnCore.ServiceLibrary.Models
{
    /// <summary>
    /// Class describing a user profile
    /// </summary>
    public class UserProfile
    {
        /// <summary>
        /// Gets or sets the ID of the user
        /// </summary>
        public int UserId { get; set; }

        /// <summary>
        /// Gets or sets the username
        /// </summary>
        public string UserName { get; set; }

        /// <summary>
        /// Gets or sets the phone number
        /// </summary>
        public string PhoneNumber { get; set; }

        /// <summary>
        /// Gets or sets the email address
        /// </summary>
        public string Email { get; set; }

        /// <summary>
        /// Gets or sets the party ID
        /// </summary>
        public int PartyId { get; set; }

        /// <summary>
        /// Gets or sets the <see cref="Party"/>
        /// </summary>
        public Party Party { get; set; }

        /// <summary>
        /// Gets or sets the <see cref="UserType"/>
        /// </summary>
        public UserType UserType { get; set; }

        /// <summary>
        /// Gets or sets the <see cref="ProfileSettingPreference"/>
        /// </summary>
        public ProfileSettingPreference ProfileSettingPreference { get; set; }
    }
}

File: src/App/frontend/src/http-client/api-client/queries/userProfile.ts
import { queryOptions, useQuery, useQueryClient } from '@tanstack/react-query';

import { apiClient } from '../client';

// ============================================================
// Types
// ============================================================

export type UserProfileResponse = {
  userId: number;
  userName: string;
  phoneNumber?: string;
  email?: string;
  partyId: number;
  party?: {
    partyId: number;
    partyTypeName: number;
    orgNumber?: string;
    ssn?: string;
    name: string;
  };
  userType: number;
  profileSettingPreference?: {
    language: string;
    preSelectedPartyId: number;
    doNotPromptForParty: boolean;
  };
};

// ============================================================
// Query Key
// ============================================================

export const userProfileKeys = {
  all: ['userProfile'] as const,
};

// ============================================================
// Pure HTTP Layer
// ============================================================

export async function fetchUserProfile(): Promise<UserProfileResponse> {
  const url = '/api/v1/profile/user';
  const response = await apiClient.get<UserProfileResponse>(url);
  return response.data;
}

// ============================================================
// Query Options
// ============================================================

export function userProfileQueryOptions() {
  return queryOptions({
    queryKey: userProfileKeys.all,
    queryFn: fetchUserProfile,
  });
}

// ============================================================
// Hooks
// ============================================================

/** Full React Query hook */
export function useUserProfileQuery() {
  return useQuery(userProfileQueryOptions());
}

/** Simple data hook */
export function useUserProfile() {
  const { data } = useUserProfileQuery();
  return data;
}

/** Invalidation hook */
export function useInvalidateUserProfile() {
  const queryClient = useQueryClient();

  return () =>
    queryClient.invalidateQueries({
      queryKey: userProfileKeys.all,
    });
}
